<div class="header">
  <ol class="breadcrumb">
    <li><a href="#/{{ apiPath }}/studies">Studies</a></li>
    <li><a href="#/{{ apiPath }}/samples/{{ seq.sample.id }}">{{ seq.sample.name }}</a></li>
    <li>Sequence <i>{{ seq_id }}</i></li>
  </ol>
</div>
<div class='notifications top-right'></div>
<button type="button" class="btn btn-success pull-right"
    ng-click="addPin()">
  <i class="fa fa-thumb-tack fa-fw"></i> Pin Page
</button>

<div class="alert alert-danger" role="alert" ng-show="seq.levenshtein_dist">
  <p><strong>This sequence has been flagged as having a possible indel or bad
  alignment </strong></p>
  <p>This occurs when at least one of the following are true:
  <ul>
    <li>The V match percentage is less than 60%</li>
    <li>The sequence was shorter than the germline after alignment, indicating a
    poor J alignment</li>
  </ul>
  </p>
  <p>
  When this occurs, the <a
  href="http://en.wikipedia.org/wiki/Levenshtein_distance" class="alert-link"
  target="_blank">Levenshtein distance</a> is calculated.  The value for this
  sequence is <strong>{{ seq.levenshtein_dist }}</strong>, indicating the
  sequence is <strong>{{ 100 * (seq.germline.length - seq.levenshtein_dist)/
  seq.germline.length | number:2 }}%</strong> similar to the germline.</p>
</div>
<h3>Sequence Details
    <span class="faded" ng-if="seq_id!=seq.seq_id">
        (Alias of {{ seq.seq_id }})
    </span>
</h3>
<div class="table-responsive">
  <table class="table table-bordered">
    <tr class="active">
      <th colspan="4" class="text-center">General</th>
    </tr>
    <tr>
      <th>
        Name
        <a data-toggle="tooltip" title="A sample-unique identifier for this
        sequence"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq_id }}</th>
      <th>
        Sample
        <a data-toggle="tooltip" title="The sample which contains this
        sequence"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td><a href="#/{{ apiPath }}/samples/{{ seq.sample.id }}" target="_blank">{{
        seq.sample.name }}</a> (#{{ seq.sample.id }})</td>
    </tr>

    <tr>
      <th>
        In-Frame
        <a data-toggle="tooltip" title="If the sequence's length is divisible by
        3"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.in_frame ? 'Yes' : 'No' }}</td>
      <th>
        Stop Codon
        <a data-toggle="tooltip" title="If the sequence's contained a stop
        codon"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.stop ? 'Yes' : 'No' }}</td>
    </tr>

    <tr>
      <th>
        Functional
        <a data-toggle="tooltip" title="If the sequence is in-frame and there is
        no stop codon"><span class="glyphicon
        glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.functional ? 'Yes' : 'No' }}</td>
      <th>
      Alignment Method
      <a data-toggle="tooltip" title="'R1+R2' if the sequence is an aligned
      bi-directional read, otherwise 'R1' or 'R2'"><span class="glyphicon
      glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.alignment }}</td>
    </tr>

    <tr>
      <th>
        Copy Number in Sample
        <a data-toggle="tooltip" title="Copies of this sequence (after filling
        in germline) that occured in the sample"><span class="glyphicon
        glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.copy_number }}</td>
      <th>
        Total Copy Number
        <a data-toggle="tooltip" title="Copies of this sequence (after filling
        in germline) that occured anywhere"><span class="glyphicon
        glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.total_copy_number }}</td>
    </tr>


    <tr class="active">
      <th colspan="4" class="text-center">Sample</th>
    </tr>

    <tr>
      <th>Subject</th>
      <td><a href="#/{{ apiPath }}/subject/{{ seq.sample.subject.id }}">{{ seq.sample.subject.identifier
      }}</a></td>
      <th>Subset</th>
      <td>{{ seq.sample.subset ? seq.sample.subset : 'N/A' }}</td>
    </tr>

    <tr>
      <th>Disease</th>
      <td>{{ seq.sample.disease }}</td>
      <th>Tissue</th>
      <td>{{ seq.sample.tissue ? seq.sample.tissue : 'N/A' }}</td>
    </tr>

    <tr class="active">
      <th colspan="4" class="text-center">Experiment</th>
    </tr>

    <tr>
      <th>Date</th>
      <td>{{ seq.sample.date }}</td>
      <th>Lab</th>
      <td>{{ seq.sample.lab }}</td>
    </tr>

    <tr>
      <th>Experimenter</th>
      <td>{{ seq.sample.experimenter }}</td>
    </tr>

    <tr class="active">
      <th colspan="4" class="text-center">CDR3</th>
    </tr>

    <tr>
      <th>
        NT Sequence
        <a data-toggle="tooltip" title="The CDR3 nucleotides"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td colspan="3" ng-bind-html="seq.junction_nt | dna"></td>
    </tr>
    <tr>
      <th>
        AA Sequence
        <a data-toggle="tooltip" title="The CDR3 amino acids"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td ng-bind-html="seq.junction_aa | aminoAcid"></td>
      <th>
      Length
        <a data-toggle="tooltip" title="Number of nucleotides in the CDR3"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.junction_nt.length }}</td>
    </tr>

    <tr class="active">
      <th colspan="4" class="text-center">V Gene</th>
    </tr>

    <tr>
      <th>
        Gene
        <a data-toggle="tooltip" title="The sequence's V gene"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td colspan="3">{{ seq.v_gene }}</td>
    </tr>

    <tr class="info">
      <th colspan="4" class="text-center">Excluding Gaps</th>
    </tr>

    <tr>
      <th>
        Total Length
        <a data-toggle="tooltip" title="The sequence's V gene length excluding
        gaps"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.v_length }}</td>
      <th>
        Total Match
        <a data-toggle="tooltip" title="The number of V gene bases matching the
        germline"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.v_match }} <span class="faded">({{ 100 * seq.v_match / seq.v_length | number:1 }}%)</span></td>
    </tr>

    <tr>
      <th>
        Pre-CDR3 Length
        <a data-toggle="tooltip" title="The number of V gene bases
        before the CDR3 excluding gaps"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.pre_cdr3_length }}</td>
      <th>
        Pre-CDR3 Match
        <a data-toggle="tooltip" title="The number of V gene bases matching the
        germline before the CDR3"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.pre_cdr3_match }} <span class="faded">({{ 100 *
      seq.pre_cdr3_match / seq.pre_cdr3_length | number:1 }}%)</span></td>
    </tr>


    <tr class="info">
      <th colspan="4" class="text-center">Including Gaps</th>
    </tr>
    <tr>
      <th>
        Total Length
        <a data-toggle="tooltip" title="The sequence's V gene length including
        gaps"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.v_length + seq.num_gaps }}</td>
      <th>
        Pre-CDR3 Length
        <a data-toggle="tooltip" title="The number of V gene bases
        before the CDR3 including gaps"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.pre_cdr3_length + seq.num_gaps }}</td>
    </tr>

    <tr class="active">
      <th colspan="4" class="text-center">J Gene</th>
    </tr>

    <tr>
      <th>
        Gene
        <a data-toggle="tooltip" title="The sequence's J gene"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td colspan="3">{{ seq.j_gene }}</td>

    <tr>
      <th>
        Total Length
        <a data-toggle="tooltip" title="The sequence's J gene length"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.j_length }}</td>
      <th>
        Total Match
        <a data-toggle="tooltip" title="The number of J gene bases matching the
        germline"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.j_match }} <span class="faded">({{ 100 * seq.j_match / seq.j_length | number:1 }}%)</span></td>

    <tr>
      <th>
        Post-CDR3 Length
        <a data-toggle="tooltip" title="The number of J gene bases
        after the CDR3"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.post_cdr3_length }}</td>
      <th>
        Post-CDR3 Match
        <a data-toggle="tooltip" title="The number of J gene bases matching the
        germline after the CDR3"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td>{{ seq.post_cdr3_match }} <span class="faded">({{ 100 *
      seq.post_cdr3_match / seq.post_cdr3_length | number:1 }}%)</span></td>
    </tr>

  </table>
</div>
<h3>Sequence View</h3>
<div class="slideWrap">
  <span class="scroller">
      <canvas id="germline-compare">
      Your browser doesn't support HTML5! Please download a newer browser.
      </canvas>
  </span>
</div>

<h3>Clone Information</h3>
<div class="table-responsive">
  <table class="table table-bordered" ng-show="!seq.clone">
    <tr>
      <td class="text-center"><strong>This sequence does not belong to a clone</strong></td>
    </tr>
  </table>
  <table class="table table-bordered" ng-show="seq.clone">
    <tr class="active">
      <th colspan="4" class="text-center">General</th>
    </tr>
    <tr>
      <th>
        Clone ID
        <a data-toggle="tooltip" title="The clone ID for this clone in the
        currently selected database version"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td><a href="#/{{ apiPath }}/clone_compare/{{ seq.clone.id }}" target="_blank">{{
        seq.clone.id }}</a></td>
      <th>
        Clone Group ID
        <a data-toggle="tooltip" title="The group of the clone as defined by V, J,
        CDR3 AAs, and subject. This can be useful for finding similar clones across
        database versions"><span class="glyphicon glyphicon-question-sign"></span></a>
      </th>
      <td><a href="#/{{ apiPath }}/clones/?group={{ seq.clone.group.id }}"
        target="_blank">{{ seq.clone.group.id }}</a></td>
    </tr>
    <tr>
      <th>
        Consensus CDR3 Nucleotides
      </th>
      <td ng-bind-html="seq.clone.cdr3_nt | dna" colspan="3"></td>
    </tr>
    <tr>
      <th>
        Consensus CDR3 Amino Acids 
      </th>
      <td ng-bind-html="seq.clone.group.cdr3_aa | aminoAcid" colspan="3"></td>
    </tr>
  </table>
</div>

<h3>Possible Duplicate Sequences
<small>All other reads sharing the same germline-filled sequence</small>
</h3>
<div class="table-responsive">
  <table class="table table-bordered" ng-show="seq.duplicates.length == 0">
    <tr>
      <td class="text-center"><strong>There are no sequences in the sample which
      have a different read length, but are identical to this one</strong></td>
    </tr>
  </table>
  <table class="table table-bordered" ng-show="seq.duplicates.length > 0">
    <thead>
      <tr>
        <th>Sample</th>
        <th>Sequence ID</th>
        <th>Alignment Method</th>
        <th>Copy Number</th>
      </tr>
    </thead>
    <tbody>
      <tr ng-repeat="dup in seq.duplicates">
        <td><a href="#/{{ apiPath }}/samples/{{ dup.sample.id }}" target="_blank">{{ dup.sample.name }}</a> (#{{ dup.sample.id }})</td>
        <td><a href="#/{{ apiPath }}/sequence/{{ dup.sample.id }}/{{ dup.seq_id }}" target="_blank">{{ dup.seq_id }}</a></td>
        <td>{{ dup.alignment }}</td>
        <td>{{ dup.copy_number }}</td>
      </tr>
    </tbody>
  </table>
</div>
